### DISCUSSION TITLE
cumulative patch and klp_patch->replace

### DISCUSSION THREAD

**Discussion by liu-song-6** [dynup/kpatch#1182] [Q&A] [answered]
Hi, 

I am building a live patch management framework. 
Based on my understanding and experience, cumulative patch would simplify the work and is highly recommended. From kernel livepatch interface, klp_patch->replace can be used to do atomic switch from older version of patches to newer version. However, IIUC, kpatch does not use the replace flag. I wonder whether this is a design choice, or there is work in progress. I think the replace flag is still useful in some cases. 

Thanks,
Song

**Comment by liu-song-6** [comment]
After some more thoughts, I think the replace feature will still be useful. So I created pull request #1183. 

Thanks,
Song

**Comment by joe-lawrence** [comment]
Hi Song,

Good question.  For us at Red Hat, I don't think we've been in a hurry to add atomic replace support to kpatch since we are still building and supporting older rhel-7 kpatches where atomic replace isn't supported.  That said, I think it would be good for kpatch-build to support newer kernel features like this.  Another one that comes to mind is the [livepatch system state API](https://lore.kernel.org/live-patching/20191030154313.13263-1-pmladek@suse.com/).

For coding specifics, I'll add notes in a review for PR #1183

**Comment by joe-lawrence** [comment, marked as answer]
Implemented (thanks @liu-song-6 !) by: https://github.com/dynup/kpatch/pull/1205 and https://github.com/dynup/kpatch/pull/1183
Tagged in v0.9.4 by: https://github.com/dynup/kpatch/pull/1215
